tests/test_mesh.py

```diff
@@ -1002,6 +1002,13 @@ def test_tag_cell_fracs():
         assert_equal(m.cell_largest_frac[i],
                      exp_cell_largest_frac[i])
 
+def test_tag_e_bins():
+    m = gen_mesh()
+    e_bins = np.array([0.0, 0.1, 20])
+
+    m.tag_e_bins(e_bins)
```

Comment:
This test looks good to me. Based on my comment above we may want to create the tag manually here instead.

Reply:
@pshriwise I have moved the `tag_e_bins` to `r2s` part as it is used in `r2s step2`. The test function fails for me (with pymoab version '5.1.0'), here is the error message:
```
FAIL: test_r2s.test_tag_e_bins
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gw/.local/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/gw/opt/pyne/tests/test_r2s.py", line 378, in test_tag_e_bins
    assert_array_equal(m.e_bins[:], e_bins)
  File "/home/gw/.local/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 931, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/home/gw/.local/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 759, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

(shapes (4, 3), (3,) mismatch)
 x: array([[ 0. ,  0.1, 20. ],
       [ 0. ,  0.1, 20. ],
       [ 0. ,  0.1, 20. ],
       [ 0. ,  0.1, 20. ]])
 y: array([ 0. ,  0.1, 20. ])
```
How should we do it if we want to create the tag `manually`?

Reply:
Sorry for taking to long to get back to you @zxkjack123. It looks like there are a couple of things preventing the behavior you want. First, there's this behavior at the bottom of `Mesh.tag`:

https://github.com/pyne/pyne/blob/dc5fccf9b1624fc6e70e976f6c433e5a2e7f074b/pyne/mesh.py#L1020-L1021

If a tag is added to the mesh with the `value` keyword set. The method will set that value for all elements on the mesh. This is happening in the `t[:] = value` line. So even though it is being stored in a sparse manner in PyMOAB, all elements are getting data and it behaves like a dense tag.

Second, there's a similar line in the `NativeMeshTag` constructor:

https://github.com/pyne/pyne/blob/dc5fccf9b1624fc6e70e976f6c433e5a2e7f074b/pyne/mesh.py#L398-L399

which will set a default value if provided. Changing the default value to `None` should fix this problem, though if a tag is dense it's considered best practice to provide a default value. I'll leave it up to you as to how you'd like to address that.

With that fix to the second problem in place, the best way I can see to set a single tag value for the mesh in `tag_e_bins` would be

```python
def tag(m, e_bins, tag_name='e_bins'):
    m.tag(name=tag_name, doc='energy boundaries of the photon source',
              tagtype=NativeMeshTag, size=len(e_bins), dtype=float,          
              storage_type='sparse')                                                                                                                                                                                                                    
    m.get_tag(tag_name)[m] = e_bxins
    return m
```

This syntax tags the energy bins on the mesh instance.

The `test_tag_e_bins` function will need to be updated as well to only check for the energy bins on the mesh instance:

```python
assert_array_equal(m.e_bins[m], e_bins)
```

I could see that becoming a one-liner function in `r2s.py` for convenience:

```python
def get_e_bins(m):
    return m.e_bins[m]
```